src/main/java/build/buildfarm/cas/cfc/CASFileCache.java

```diff
@@ -744,16 +744,26 @@ public void write(byte[] b, int off, int len) throws IOException {
 
     @Override
     public void close() throws IOException {
-      // we ignore closes below the complete size
-      if (out.getWritten() >= size) {
-        super.close();
-      }
+      // disallow any further writes
       closed = true;
-      onClosed.run();
+      try {
+        // we ignore closes below the complete size
+        if (out.getWritten() < size) {
+          // onClosed responsibility lies with delegate stream
+          onClosed.run();
+        } else {
+          super.close();
```

Comment:
there is no onClose.run() triggered in this block, is it as expected ?

Reply:
It is. The parent class is expected to invoke onClosed.run(), and does so at https://github.com/bazelbuild/bazel-buildfarm/blob/3c1562d137afb1181320b86b3f1c826955d93dbb/src/main/java/build/buildfarm/cas/cfc/CASFileCache.java#L1019

I have mixed feelings about most of this - the exposure of onClosed.run() to the newOutput actually seems like it could be avoided. But thanks for taking a close look at this.

Reply:
got it, thanks!

Reply:
I've reworked this and included some more testing. The onClosed now indicates whether it has been cancelled, and while this is not a public API, I wanted those behaviors to have some consistency. Tests there ensure that Write::newOutput handles closes nicely. More testing would be required to ensure that we block as expected.